src/process/tracking.rs

```diff
@@ -0,0 +1,312 @@
+    /// The index in the event series.
+    pub(crate) idx: u32,
+}
+
+#[allow(dead_code)]
+impl TrackingInfo {
+    /// Get the tracking id.
+    pub(crate) fn tracking_id(&self) -> u128 {
+        (self.first_ts as u128) << 64 | self.orig_head as u128
+    }
+}
+
+impl Eq for TrackingInfo {}
+
+impl PartialEq for TrackingInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.first_ts.eq(&other.first_ts) && self.orig_head.eq(&other.orig_head)
+    }
+}
+
+impl PartialOrd for TrackingInfo {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Ord for TrackingInfo {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.first_ts
+            .cmp(&other.first_ts)
+            .then_with(|| self.orig_head.cmp(&other.orig_head))
+    }
+}
+
+#[derive(Default)]
+#[event_section_factory(TrackingInfo)]
+pub(crate) struct TrackingInfoEventFactory {}
+
+impl RawEventSectionFactory for TrackingInfoEventFactory {
+    fn from_raw(&mut self, _: Vec<BpfRawSection>) -> Result<Box<dyn EventSection>> {
+        bail!("TrackingInfoEvents cannot be created from bpf")
+    }
+}
+
+impl TrackingInfo {
+    fn new(track: &SkbTrackingEvent) -> Result<Self> {
+        Ok(TrackingInfo {
+            orig_head: track.orig_head,
+            first_ts: track.timestamp,
+            idx: 0,
+        })
+    }
+}
+
+// Data identifying an OvsUpcall Event
+#[derive(Debug, PartialEq, Eq, Hash)]
+struct UpcallKey {
+    /// Cpu of the upcall event
+    cpu: u32,
+    /// Timestamp of the upcall event
+    ts: u64,
+}
+
+/// AddTracking is a ProcessorAction that looks at the events' tracking information and inserts
+/// information about the previous event of the same series.
+pub(crate) struct AddTracking {
+    /// Skb tracking map. Indexed by skb tracking id, this map used to access the TrackingInfo for
+    /// each tracking id. Also, it allows us to "overwrite" the tracking information of certain
+    /// events.
+    skb_tracking: HashMap<u128, Arc<Mutex<TrackingInfo>>>,
+    /// When an ovs queue_id is generated, it's skb tracking info is stored in this map. The
+    /// queue_id of future ovs events are looked up in this table to get the right tracking
+    /// information.
+    ovs_queue_tracking: HashMap<u32, Arc<Mutex<TrackingInfo>>>,
+    /// When an upcall happens, the packet might get fragmented. This map is used to use the same
+    /// TrackingInfo for all fragments.
+    ovs_upcalls_tracking: HashMap<UpcallKey, Arc<Mutex<TrackingInfo>>>,
+}
+
+impl AddTracking {
+    pub(crate) fn new() -> Self {
+        AddTracking {
+            skb_tracking: HashMap::new(),
+            ovs_queue_tracking: HashMap::new(),
+            ovs_upcalls_tracking: HashMap::new(),
+        }
+    }
+
+    /// Process one event adding TrackingInfo section.
+    pub(crate) fn process_one(&mut self, event: &mut Event) -> Result<()> {
+        if let Some(ovs) = event.get_section::<OvsEvent>(ModuleId::Ovs) {
```

Comment:
We don't call `process_skb` for all events:
- If an action event contains a queue_id we don't want to call `process_skb` before because it will create a new entry in the tracking map for an skb (the newly created) that we don't want to track directly (we want to use the tracking info carried along from ovs).
- UpcallEnqueue events have skb pointers that point to fragments of the overall skb that is being tracked so we don't want to call `process_skb` on that one either IIUC.